### PR TITLE
First tiny step towards fully OTF upgrade sims 

### PR DESCRIPTION
**Pull request by ddobrigk**

### CODE REVIEW

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+      Float_t lPosition = lX0 + (lX1 - lX0) * ((Float_t)(iStep)) / ((Float_t)(lNStepsLIntegrator - 1));
+      track.propagateTo(lPosition, lMagneticField);
+      track.getXYZGlo(lPointNplus);
+      lLength += std::hypot(lPointNplus[0] - lPointN[0], lPointNplus[1] - lPointN[1], lPointNplus[2] - lPointN[2]);
+    }
+    return lLength;
+  }
+
+  Double_t Velocity(Double_t lMomentum, Double_t lMass)
+  {
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:163]: Same comments as for `TrackLength`.

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    // Momentum p and mass m -> returns speed in centimeters per picosecond
+    // Useful for TOF calculations
+    Double_t lA = TMath::Power(lMomentum / lMass, 2);
+    return 0.0299792458 * TMath::Sqrt(lA / (1 + lA));
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:167]: The 0.0299792458 constant needs some documentation.

Comment (vkucera) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:167]: I would suggest using `std::` functions.

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  void process(soa::Join<aod::Collisions, aod::McCollisionLabels>::iterator const& collision, soa::Join<aod::Tracks, aod::TracksCov, aod::McTrackLabels> const& tracks, aod::McParticles const&)
+  {
+    o2::math_utils::Point3D<float> pvpos{collision.posX(), collision.posY(), collision.posZ()};
+    std::array<float, 6> pvcov;
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:173]:
```suggestion
    o2::math_utils::Point3D<float> pvPos{collision.posX(), collision.posY(), collision.posZ()};
    std::array<float, 6> pvCov;
```
naming conventions

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    o2::dataformats::VertexBase pvvtx(pvpos, pvcov);
+
+    auto mcCollision = collision.mcCollision();
+    o2::math_utils::Point3D<float> mcpvpos{mcCollision.posX(), mcCollision.posY(), mcCollision.posZ()};
+    std::array<float, 6> mcpvcov; // dummy!
+    for (int ii = 0; ii < 6; ii++)
+      mcpvcov[ii] = 1e-6;
+    o2::dataformats::VertexBase mcpvvtx(mcpvpos, mcpvcov);
+
+    for (auto& track : tracks) {
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:189]:
```suggestion
    for (const auto& track : tracks) {
```

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+      // first step: find precise arrival time (if any)
+      // --- convert track into perfect track
+      if (!track.has_mcParticle()) // should always be OK but check please
+        LOG(error) << "Oh no! No mcParticle label for this track! This shouldn't happen!";
+
+      o2::track::TrackParCov o2track;
+      auto mcParticle = track.mcParticle();
+      convertMCParticleToO2Track(mcParticle, o2track);
+
+      float lX_PV = -100, lX_iTOF = -100, lX_oTOF = -100, lThisTrackLength_iTOF = -1, lThisTrackLength_oTOF = -1;
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:199]: The variables names should be improved and underscores removed.

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+      upgradetof(lNSigma_iTOF[0], lNSigma_iTOF[1], lNSigma_iTOF[2], lNSigma_iTOF[3], lNSigma_iTOF[4],
+                 lNSigma_oTOF[0], lNSigma_oTOF[1], lNSigma_oTOF[2], lNSigma_oTOF[3], lNSigma_oTOF[4]);
+    }
+  }
+};
+
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  WorkflowSpec workflow{adaptAnalysisTask<onTheFlyTOFPID>(cfgc)};
+  return workflow;
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:278]: This can be one line: `return WorkflowSpec{...`

**Review by vkucera — CHANGES_REQUESTED** [review]
Hi @ddobrigk , I added some suggestions for improvement.

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+//
+// Task to add a table of track parameters propagated to the primary vertex
+//
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:14]: Thanks, done in local copy (commit to follow)

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+// More particles could be added but would have to be added to the LUT
+// being used in the onTheFly tracker task.
+
+using namespace o2;
+using namespace o2::framework;
+
+namespace o2::aod
+{
+namespace upgradetof
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:51]: Done (even if our own main data model doesn't follow this convention on a very large amount of instances but ok)

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+  Produces<aod::UpgradeTof> upgradetof;
+
+  // necessary for particle charges
+  Service<O2DatabasePDG> pdg;
+
+  // these are the settings governing the TOF layers to be used
+  // note that there are two layers foreseen for now: inner and outer TOF
+  // more could be added (especially a disk TOF at a certain z?)
+  // in the evolution of this effort
+  Configurable<float> dBz{"dBz", 20, "magnetic field (kilogauss)"};
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:87]: I would leave as is for the time being. An important part of the ALICE 3 scoping tests is exactly to have the magnetic field fully configurable and not actually fixed, so this is actually a feature, not a bug :-D

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 2 above)
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:167]: done!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 3 above)
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:173]: done!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 5 above)
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:199]: I would prefer to keep the underscores, and the names are as informative as they should be in the context of the code. For instance, the first variables correspond to Track X at inner and at outer TOF, and the underscore kind of serves exactly to highlight that. I considered doing "XAtiTOF" but it's just confusing without the underscores...

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 6 above)
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:278]: done!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+#include "Common/Core/trackUtilities.h"
+#include "ReconstructionDataFormats/DCA.h"
+#include "DetectorsBase/Propagator.h"
+#include "DetectorsBase/GeometryManager.h"
+#include "CommonUtils/NameConf.h"
+#include "CCDB/CcdbApi.h"
+#include "DataFormatsParameters/GRPMagField.h"
+#include "CCDB/BasicCCDBManager.h"
+#include "Framework/HistogramRegistry.h"
+#include "Framework/runDataProcessing.h"
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:30]: fixed/adjusted!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 4 above)
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:189]: easy to forget :-) added!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+
+  // these are the settings governing the TOF layers to be used
+  // note that there are two layers foreseen for now: inner and outer TOF
+  // more could be added (especially a disk TOF at a certain z?)
+  // in the evolution of this effort
+  Configurable<float> dBz{"dBz", 20, "magnetic field (kilogauss)"};
+  Configurable<float> iTOFRadius{"iTOFRadius", 20, "barrel iTOF radius (cm)"};
+  Configurable<float> oTOFRadius{"oTOFRadius", 80, "barrel oTOF radius (cm)"};
+  Configurable<float> iTOFTimeReso{"iTOFTimeReso", 20, "barrel iTOF time error (ps)"};
+  Configurable<float> oTOFTimeReso{"oTOFTimeReso", 20, "barrel oTOF time error (ps)"};
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:91]: This is not at all "quite cryptic" for ALICE 3 ;-) - I would say it's as cryptic as any other detector acronym - but granted, I can add a statement

[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:91]: Actually, a few lines above this I explicitly say "inner and outer TOF" ;-)

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    }
+  }
+
+  template <typename mcParticleType>
+  void convertMCParticleToO2Track(mcParticleType& particle, o2::track::TrackParCov& o2track)
+  {
+    // FIXME: this is a fundamentally important piece of code.
+    // It could be placed in a utility file instead of here.
+    std::array<float, 3> xyz = {static_cast<float>(particle.vx()), static_cast<float>(particle.vy()), static_cast<float>(particle.vz())};
+    std::array<float, 3> ptetaphi = {static_cast<float>(particle.pt()), static_cast<float>(particle.eta()), static_cast<float>(particle.phi())};
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:121]: The benefit is readability, which is actually quite important.  I prefer to keep as is

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+    auto pdgInfo = pdg->GetParticle(particle.pdgCode());
+    int charge = 0;
+    if (pdgInfo != nullptr)
+      charge = pdgInfo->Charge();
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:125]: I don't see the point of adding extra lines to the code for brackets for a single-line if clause. It makes the code unnecessarily long, which impacts readability. I prefer to keep as is

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+    std::array<float, 5> params;
+    std::array<float, 15> covm = {
+      0.,
+      0., 0.,
+      0., 0., 0.,
+      0., 0., 0., 0.,
+      0., 0., 0., 0., 0.};
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:132]: done!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+    float s, c, x;
+    o2::math_utils::sincos(ptetaphi[2], s, c);
+    o2::math_utils::rotateZInv(xyz[0], xyz[1], x, params[0], s, c);
+    params[1] = xyz[2];
+    params[2] = 0.; // since alpha = phi
+    auto theta = 2. * atan(exp(-ptetaphi[1]));
+    params[3] = 1. / tan(theta);
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:139]: done!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+    params[4] = charge / ptetaphi[0];
+
+    // Initialize TrackParCov in-place
+    new (&o2track)(o2::track::TrackParCov)(x, ptetaphi[2], params, covm);
+  }
+
+  Double_t TrackLength(o2::track::TrackParCov track, Double_t lX0, Double_t lX1, Double_t lMagneticField)
+  {
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:147]: done!

[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:147]: changed

[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:147]: added

**Comment by ddobrigk** [comment]
@vkucera thank you! Can you please check if you are satisfied and approve if all is good? Thanks and have a nice evening!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+      track.getXYZGlo(lPointN);
+      float lPosition = lX0 + (lX1 - lX0) * (static_cast<float>(iStep)) / (static_cast<float>(lNStepsLIntegrator - 1));
+      track.propagateTo(lPosition, lMagneticField);
+      track.getXYZGlo(lPointNplus);
+      lLength += std::hypot(lPointNplus[0] - lPointN[0], lPointNplus[1] - lPointN[1], lPointNplus[2] - lPointN[2]);
+    }
+    return lLength;
+  }
+
+  float Velocity(float lMomentum, float lMass)
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:158]: Same comment about the `l`.

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+  {
+    // Momentum p and mass m -> returns speed in centimeters per picosecond
+    // Useful for TOF calculations
+    float lA = std::pow(lMomentum / lMass, 2);
+    return (o2::constants::physics::LightSpeedCm2NS / 1e+3) * TMath::Sqrt(lA / (1 + lA));
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:163]:
```suggestion
    return (o2::constants::physics::LightSpeedCm2NS / 1e+3) * std::sqrt(lA / (1 + lA));
```

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+  {
+    o2::math_utils::Point3D<float> pvPos{collision.posX(), collision.posY(), collision.posZ()};
+    std::array<float, 6> pvCov;
+    pvCov[0] = collision.covXX();
+    pvCov[1] = collision.covXY();
+    pvCov[2] = collision.covYY();
+    pvCov[3] = collision.covXZ();
+    pvCov[4] = collision.covXY();
+    pvCov[5] = collision.covXZ();
+    o2::dataformats::VertexBase pvvtx(pvPos, pvCov);
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:176]:
```suggestion
    o2::dataformats::VertexBase pvVtx(pvPos, pvCov);
```

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+
+    auto mcCollision = collision.mcCollision();
+    o2::math_utils::Point3D<float> mcPvPos{mcCollision.posX(), mcCollision.posY(), mcCollision.posZ()};
+    std::array<float, 6> mcPvCov; // dummy!
+    for (int ii = 0; ii < 6; ii++)
+      mcPvCov[ii] = 1e-6;
+    o2::dataformats::VertexBase mcpvvtx(mcPvPos, mcPvCov);
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:183]:
```suggestion
    o2::dataformats::VertexBase mcPvVtx(mcPvPos, mcPvCov);
```

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 5 above)
Comment (vkucera) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:195]: Underscores are allowed only in the names of namespaces. Names of variables follow the [camelCase convention](https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=Variable_Names#Variable_Names).
I think that `xPv`, `xITof` (`xITOF`), `xOTof` (`xOTOF`) are pretty clear.

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 21]
```diff
@@ -0,0 +1,274 @@
...
+        lX_oTOF = -100;
+      if (lX_PV > -99. && lX_iTOF > -99.)
+        lThisTrackLengthReco_iTOF = trackLength(recoTrack, lX_PV, lX_iTOF, dBz);
+      if (lX_PV > -99. && lX_oTOF > -99.)
+        lThisTrackLengthReco_oTOF = trackLength(recoTrack, lX_PV, lX_oTOF, dBz);
+
+      // Straight to Nsigma
+      float lDeltaTime_iTOF[5], lNSigma_iTOF[5];
+      float lDeltaTime_oTOF[5], lNSigma_oTOF[5];
+      int lpdg_array[5] = {11, 13, 211, 321, 2212};
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:234]: Don't use [magic numbers](https://rawgit.com/AliceO2Group/CodingGuidelines/master/coding_guidelines.html?showone=Magic_numbers#Magic_numbers). Use the enum particle names.

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 22]
```diff
@@ -0,0 +1,274 @@
...
+      float lMasses[5];
+
+      if (doQAplots) {
+        float momentum = recoTrack.getP();
+        float innerbeta = (lThisTrackLength_iTOF / (1e+3 * lMeasuredTime_iTOF)) / o2::constants::physics::LightSpeedCm2NS;
+        float outerbeta = (lThisTrackLength_oTOF / (1e+3 * lMeasuredTime_oTOF)) / o2::constants::physics::LightSpeedCm2NS;
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:240]: Please fix the variable names.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 5 above)
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:195]: OK! Changed, thanks!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+    if (doQAplots) {
+      const AxisSpec axisMomentum{static_cast<int>(80), 0.0f, +4.0f, "#it{p} (GeV/#it{c})"};
+      const AxisSpec axisVelocity{static_cast<int>(110), 0.0f, +1.1f, "Measured #beta"};
+      histos.add("h2dVelocityVsMomentumInner", "h2dVelocityVsMomentumInner", kTH2F, {axisMomentum, axisVelocity});
+      histos.add("h2dVelocityVsMomentumOuter", "h2dVelocityVsMomentumOuter", kTH2F, {axisMomentum, axisVelocity});
+    }
+  }
+
+  template <typename mcParticleType>
+  void convertMCParticleToO2Track(mcParticleType& particle, o2::track::TrackParCov& o2track)
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:115]: done!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 19 above)
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:176]: Ah yes this one I missed, thanks, done

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 24]
```diff
@@ -0,0 +1,274 @@
...
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+//
+// Task to add a table of track parameters propagated to the primary vertex
+//
```
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:14]: Implemented!

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 12 above)
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:120]: Changed! Thanks

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 13 above)
[the remark this answers is not in the document]
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:124]: Boh, it's okay, I can add braces :-D done

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx (hunk 17 above)
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTOFPID.cxx:158]: no 'l' anywhere anymore in latest version (commit soon)

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 25]
```diff
@@ -0,0 +1,280 @@
...
+      length += std::hypot(pointNplus[0] - pointN[0], pointNplus[1] - pointN[1], pointNplus[2] - pointN[2]);
+    }
+    return length;
+  }
+
+  /// returns velocity in centimeters per picoseconds
+  /// \param momentum the input track
+  /// \param x0 the initial position
+  /// \param x1 the final position
+  /// \param magneticField the magnetic field to use when propagating
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:171]: I guess this is a copy-paste typo from `trackLength`

File: ALICE3/TableProducer/onTheFlyTOFPID.cxx [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+                  upgrade_tof::NSigmaProtonInner,
+                  upgrade_tof::NSigmaElectronOuter,
+                  upgrade_tof::NSigmaMuonOuter,
+                  upgrade_tof::NSigmaPionOuter,
+                  upgrade_tof::NSigmaKaonOuter,
+                  upgrade_tof::NSigmaProtonOuter);
+} // namespace o2::aod
+
+struct OnTheFlyTOFPID {
+  Produces<aod::UpgradeTof> upgradetof;
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTOFPID.cxx:81]:
```suggestion
  Produces<aod::UpgradeTof> upgradeTof;
```

File: ALICE3/TableProducer/onTheFlyTracker.cxx [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+    auto theta = 2. * std::atan(std::exp(-particle.eta()));
+    params[3] = 1. / std::tan(theta);
+    params[4] = charge / particle.pt();
+
+    // Initialize TrackParCov in-place
+    new (&o2track)(o2::track::TrackParCov)(x, particle.phi(), params, covm);
+  }
+
+  template <typename CollType, typename TTrackPar>
+  void FillTracksPar(CollType& coll, aod::track::TrackTypeEnum trackType, TTrackPar& trackPar)
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTracker.cxx:102]:
```suggestion
  template <typename CollType, typename TTrackPar>
  void fillTracksPar(CollType& coll, aod::track::TrackTypeEnum trackType, TTrackPar& trackPar)
```
Please add at least one line of documentation above.

File: ALICE3/TableProducer/onTheFlyTracker.cxx [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  void process(aod::McCollision const& mcCollision, aod::McParticles const& mcParticles)
+  {
+    o2::dataformats::DCA dcaInfoCov;
+    o2::dataformats::VertexBase vtx;
+
+    for (const auto& mcParticle : mcParticles) {
+      auto pdg = std::abs(mcParticle.pdgCode());
+      if (pdg != 11 && pdg != 13 && pdg != 211 && pdg != 321 && pdg != 2212)
```
Comment (vkucera) [inline ALICE3/TableProducer/onTheFlyTracker.cxx:115]: Please use the enums.

**Review by vkucera — CHANGES_REQUESTED** [review]
Hi @ddobrigk , thanks for the fixes. Please see a few minor remaining things.

File: ALICE3/TableProducer/onTheFlyTracker.cxx (hunk 28 above)
Comment (vkucera) [reply ALICE3/TableProducer/onTheFlyTracker.cxx:115]: If this list gets longer, one might consider using some search method from the standard library, like `std::find` or `std::set<int> {1, 2, 3, 4}).count(my_var)`.

File: ALICE3/TableProducer/onTheFlyTracker.cxx (hunk 27 above)
Comment (ddobrigk) [reply ALICE3/TableProducer/onTheFlyTracker.cxx:102]: fixed!

**Comment by ddobrigk** [comment]
Thanks, all done!

**Review by vkucera — APPROVED** [review]

**Pull request merged**